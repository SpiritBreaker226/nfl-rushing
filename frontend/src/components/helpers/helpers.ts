import { Dispatch } from 'react'

import queryString from 'query-string'

import { PlayerQuery } from '../../types/PlayerQuery'
import { Types } from '../../types/Actions'

export interface SetParamsCallbackProps {
  params: PlayerQuery
  dispatch: Dispatch<any>
  url: string
}

const setParamsCallback = ({
  params,
  dispatch,
  url,
}: SetParamsCallbackProps) => {
  const currentURL = queryString.parseUrl(url)
  const query = { ...currentURL.query, ...params }
  const newUrl = queryString.stringifyUrl(
    {
      url: currentURL.url,
      query,
    },
    {
      skipEmptyString: true,
      skipNull: true,
    }
  )

  dispatch({ type: Types.UpdateURL, payload: { url: newUrl } })
}

export { setParamsCallback }
