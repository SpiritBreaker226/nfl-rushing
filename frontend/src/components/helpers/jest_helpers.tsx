import React, { FunctionComponent } from 'react'

import { Player } from '../../types/Players'
import { Direction } from '../../types/PlayerQuery'
import { Pagination } from '../../types/Pagination'

import { AppContext } from '../../contexts/AppContext'
import { Action } from '../../types/Actions'

export const players: Player[] = [
  {
    attributes: {
      att: 2,
      attg: '2.0',
      avg: '3.5',
      first: 0,
      first_precentage: '0.0',
      fourty_plus: 0,
      fum: 0,
      lng: 7,
      player: 'Joe Banyard',
      pos: 'RB',
      td: 0,
      team: 'JAX',
      twenty_plus: 0,
      yds: 7,
      ydsg: '7.0',
    },
  },
  {
    attributes: {
      att: 205,
      attg: '12.8',
      avg: '5.1',
      first: 49,
      first_precentage: '23.9',
      fourty_plus: 2,
      fum: 2,
      lng: '75T',
      player: 'Mark Ingram',
      pos: 'RB',
      td: 6,
      team: 'NO',
      twenty_plus: 4,
      yds: 1,
      ydsg: '65.2',
    },
  },
  {
    attributes: {
      att: 3,
      attg: '0.8',
      avg: '-0.7',
      first: 0,
      first_precentage: '0.0',
      fourty_plus: 0,
      fum: 1,
      lng: 0,
      player: 'Brett Hundley',
      pos: 'QB',
      td: 0,
      team: 'GB',
      twenty_plus: 0,
      yds: -2,
      ydsg: '-0.5',
    },
  },
]

export interface TestingState {
  players?: Player[]
  pagination?: Pagination
  urlToPlayersEndpoint?: string
  search?: string
  errorMessage?: string
  isLoading?: boolean
  sorting?: {
    fieldName: string | undefined
    dir: Direction
  }
}

export interface MakeWrapperProps {
  state: TestingState
  dispatch?: (action: Action) => void
}

export const MakeWrapper: FunctionComponent<MakeWrapperProps> = ({
  state,
  dispatch,
  children,
}) => (
  <AppContext.Provider
    value={{
      state: {
        players: [],
        pagination: {
          rows_per_page: 10,
          page_total: 0,
          current_page: '1',
        },
        urlToPlayersEndpoint: '',
        search: '',
        errorMessage: '',
        sorting: {
          fieldName: undefined,
          dir: undefined,
        },
        isLoading: false,
        ...state,
      },
      dispatch: dispatch || (() => null),
    }}
  >
    {children}
  </AppContext.Provider>
)
