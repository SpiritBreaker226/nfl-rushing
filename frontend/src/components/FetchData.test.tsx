import React from 'react'

import { waitFor, screen, render } from '@testing-library/react'

import axios from 'axios'

import { MakeWrapper } from './helpers/jest_helpers'

import FetchData from './FetchData'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('FetchData', () => {
  describe('fetch player data', () => {
    it('renders loading screen', async () => {
      render(
        <MakeWrapper state={{ isLoading: true }}>
          <FetchData />
        </MakeWrapper>
      )

      expect(screen.getByText('Loading', { exact: false })).toBeInTheDocument()
      expect(screen.getByText('Loading', { exact: false })).toMatchSnapshot()

      await waitFor(() => {})
    })

    describe('on error from the server', () => {
      it('should render an error message', async () => {
        mockedAxios.get.mockRejectedValue(new Error('fake error message'))

        render(
          <MakeWrapper
            state={{
              errorMessage: 'fake error message',
            }}
          >
            <FetchData />
          </MakeWrapper>
        )

        await waitFor(() => screen.getByText('error', { exact: false }))

        expect(screen.getByText('error', { exact: false })).toBeInTheDocument()
        expect(screen.getByText('error', { exact: false })).toMatchSnapshot()
      })
    })
  })
})
