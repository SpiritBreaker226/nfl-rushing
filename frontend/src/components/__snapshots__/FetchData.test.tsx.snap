// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchData fetch player data on error from the server should render an error message 1`] = `
<section
  class="error"
>
  fake error message
</section>
`;

exports[`FetchData fetch player data renders loading screen 1`] = `
<div
  class="loading"
>
  Loading...
</div>
`;
